packages/tizen-remote/lib/index.js

```diff
@@ -0,0 +1,764 @@
+   * @param {any} data
+   * @param {SendOptions} [opts]
+   * @returns {Promise<void>}
+   */
+  async send(data, {noConnect = false} = {}) {
+    let ws = this.#ws;
+    /** @type {string} */
+    let payload;
+    if (!this.#ws) {
+      if (noConnect) {
+        try {
+          payload = JSON.stringify(data);
+        } catch {
+          payload = '(not serializable)';
+        }
+        throw new Error(`Disconnected; cannot send message: ${payload}`);
+      }
+      this.#debug('Disconnected when attempting to send; attempting connection...');
+      // connect() assigns this.#ws if it was successful, so we
+      // don't need to do so here.
+      ws = await this.connect();
+    }
+    try {
+      payload = JSON.stringify(data);
+      this.#debug('Sending: %O', data);
+    } catch (err) {
+      const error = /** @type {TypeError} */ (err);
+      throw new TypeError(`Cannot serialize data to JSON: ${error.message}`);
+    }
+    const send = /** @type {(data: any) => Promise<void>} */ (
+      promisify(/** @type {WebSocket} */ (ws).send).bind(ws)
+    );
+    await send(payload);
+    this.emit(Event.SENT, payload);
+  }
+
+  /**
+   * Send JSON-serializable data to the Tizen web socket server.
+   *
+   * This is low-level, and you probably want something else.
+   * @template T
+   * @param {string} channel
+   * @param {any} data
+   * @param {SendOptions} [opts]
+   * @returns {Promise<T>}
+   */
+  async sendRequest(channel, data, {noConnect = false} = {}) {
+    /** @type {string} */
+    let payload;
+    /** @type {WebSocket} */
+    let ws;
+    if (this.#ws) {
+      ws = this.#ws;
+    } else {
+      if (noConnect) {
+        throw new Error(`Disconnected; cannot send message: ${data}`);
+      }
+      this.#debug('Not connected; attempting connection...');
+      ws = await this.connect();
+    }
+
+    try {
+      payload = JSON.stringify(data);
+    } catch (err) {
+      throw new Error(`Unable to serialize data to JSON: ${/** @type {Error} */ (err).message}`);
+    }
+
+    const send = /** @type {(data: any) => Promise<void>} */ (promisify(ws.send).bind(this.#ws));
+
+    this.#debug('Sending request on channel %s: %O', channel, data);
+    await send(payload);
+
+    return await new Promise((resolve, reject) => {
+      const tokenTimer = setTimeout(() => {
+        reject(new Error(`Did not receive token in ${this.#tokenTimeout}ms`));
+      }, this.#tokenTimeout);
+
+      /** @param {import('ws').RawData} data */
+      const listener = (data) => {
+        try {
+          const resData = JSON.parse(data.toString());
+          if (resData.event === channel) {
+            resolve(resData);
+          }
+        } catch {
+          // if we can't parse the data, it's not something we're interested in
+        } finally {
+          clearTimeout(tokenTimer);
+          this.#offWs(WsEvent.MESSAGE, listener);
+        }
+      };
+      this.#onWs(WsEvent.MESSAGE, listener);
```

Comment:
yeah.  so we want to `await send()` because it might reject.  if it rejects, we don't want to be listening for messages, because then we'd have to stop listening to avoid leaks.   the `new Promise` callback is executed synchronously, so essentially creating the `message` listener blocks the event loop.  the only way for this to not work would be if a message was received before `await send()` fulfills.  I don't think that's possible, but I'm not 100%.

Reply:
fwiw, I assume you know this, but just in case:

```
new Promise(resolve => {
 console.error(1);
 resolve();
});
console.error(2);
```

always prints:

```
1
2
```